### PR TITLE
Added new contributors guide

### LINKED ISSUE #316: Contributing Guidelines Updates
I want to start a thread on updating guidelines for submitting PRs to CREDIT and updating our checks and approval process. Right now @jsschreck and I tend to do many of the larger approvals, but given the growing number of developers and contributors on the project we need to more to a more formalized system for approvals and releases.

- [ ] PR template and checklist for pre-PR checks (like running pytest)
- [ ] Decide on what range of Python/PyTorch version CI checks we want to support (not too high latency but still reasonably thorough)
- [ ] What PRs need more thorough code reviews vs what can be merged with fewer checks.
- [ ] How to assign reviewers.
- [ ] Timeline for reviews, how to communicate that to potential reviewers
- [ ] Who are the approvers?

### PR DESCRIPTION
**Pull request by djgagne**
### Description
I have updated the Contributors Guide based on the feedback from the Google Doc. New guidance includes
* LLM usage guidelines
* Details on feature additions
* Details on backwards compatibility
* Making all guidance sections more CREDIT-specific.
* Converting the guide format to Markdown from rst.

### Related Issues and PRs

Closes #316 

### Checklist
- [x] I am familiar with the [contributing guidelines](https://miles-credit.readthedocs.io/en/latest/contrib.html).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date.

### CODE REVIEW

File: docs/source/contrib.md [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+
+If you have any questions, please feel free to reach out to us on [GitHub Issues](https://github.com/NCAR/miles-credit/issues) (for bugs or feature requests) or [Github Discussions](https://github.com/NCAR/miles-credit/discussions) (for more open-ended discussions about CREDIT features or strategic directions). You can also reach us by email at [miles@ucar.edu](mailto:miles@ucar.edu).
+
+## Where to start
+
+If you are interested in contributing to the core CREDIT repository, first reach out to the MILES team at [miles@ucar.edu](mailto:miles@ucar.edu) to set up an introductory meeting and discuss your interests and what would be a good focus area. 
+
+Once we have agreed on a focus area, please start issues on the [CREDIT Issues](https://github.com/NCAR/miles-credit/issues) page so that we can track progress on our internal project board and version Milestone tracker.
```
Comment (kafitzgerald) [inline docs/source/contrib.md]:
```suggestion
Once we have agreed on a focus area, please start issues on the [CREDIT Issues](https://github.com/NCAR/miles-credit/issues) page so that we can track progress on our internal project board and [version milestone tracker](https://github.com/NCAR/miles-credit/milestones). 
```

File: docs/source/contrib.md [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+
+If you are interested in working on an existing issue, please comment on the issue to let us know you are working on it. This will help us avoid duplicate work.
+
+The code for CREDIT is hosted on GitHub. If you do not have one, you will need to create a [free GitHub account](https://github.com/signup). The [GitHub Quickstart Guide](https://docs.github.com/en/get-started/start-your-journey) is a great place to get started with git and GitHub.
+
+## Reporting bugs
+
+Something not working as expected? We would love to hear about it! Please report any bugs you find by opening an issue on GitHub. 
+
+When reporting a bug, please include as much information as possible. This will help us reproduce the bug and fix it efficiently. For more information on how to write a good bug report, see this stackoverflow post on [how to make a good bug report](https://stackoverflow.com/help/minimal-reproducible-example).
```
Comment (kafitzgerald) [inline docs/source/contrib.md]:
```suggestion
When reporting a bug, please include as much information as possible. This will help us reproduce the bug and fix it efficiently. For more information on how to write a good bug report, see this Stack Overflow post on [how to make a good bug report](https://stackoverflow.com/help/minimal-reproducible-example).  
```

File: docs/source/contrib.md [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+
+Something not working as expected? We would love to hear about it! Please report any bugs you find by opening an issue on GitHub. 
+
+When reporting a bug, please include as much information as possible. This will help us reproduce the bug and fix it efficiently. For more information on how to write a good bug report, see this stackoverflow post on [how to make a good bug report](https://stackoverflow.com/help/minimal-reproducible-example).  
+To increase the helpfulness of the bug report, please include the following information:
+
+* Compute platform (e.g., Casper, Derecho, Macbook Pro, AWS VM, etc.)  
+* Whether issue is on CPU or GPU and what kind of GPU is being used.  
+* Version of PyTorch, CUDA toolkit, NCCL (if applicable), other relevant libraries.  
+*  What stage of ML process (processing, training, inference).
```
Comment (kafitzgerald) [inline docs/source/contrib.md]:
```suggestion
* What stage of ML process (processing, training, inference).
```

File: docs/source/contrib.md [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+## Reporting bugs
+
+Something not working as expected? We would love to hear about it! Please report any bugs you find by opening an issue on GitHub. 
+
+When reporting a bug, please include as much information as possible. This will help us reproduce the bug and fix it efficiently. For more information on how to write a good bug report, see this stackoverflow post on [how to make a good bug report](https://stackoverflow.com/help/minimal-reproducible-example).  
+To increase the helpfulness of the bug report, please include the following information:
+
+* Compute platform (e.g., Casper, Derecho, Macbook Pro, AWS VM, etc.)  
+* Whether issue is on CPU or GPU and what kind of GPU is being used.  
+* Version of PyTorch, CUDA toolkit, NCCL (if applicable), other relevant libraries.
```
Comment (kafitzgerald) [inline docs/source/contrib.md:28]: If we tell folks how to do this, it might be more likely to happen.

Maybe this is a new issue though and/or belongs in an issue template instead.

File: docs/source/contrib.md [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+2. Open a pull request.  
+3. Address feedback.  
+4. Wait for your pull request to be merged.  
+5. Delete your branch.
+
+## Get set up to develop on your local machine
+
+### Internal developers and close collaborators: Clone and create a branch
+
+CREDIT developers at NCAR and close collaborators (e.g., interns, visitors, partners on funded research projects) should manage their code contributions through branches in the `NCAR/miles-credit` Github repository.
```
Comment (kafitzgerald) [inline docs/source/contrib.md]:
```suggestion
CREDIT developers at NCAR and close collaborators (e.g., interns, visitors, partners on funded research projects) should manage their code contributions through branches in the `NCAR/miles-credit` GitHub repository. 
```

File: docs/source/contrib.md [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+
+## Get set up to develop on your local machine
+
+### Internal developers and close collaborators: Clone and create a branch
+
+CREDIT developers at NCAR and close collaborators (e.g., interns, visitors, partners on funded research projects) should manage their code contributions through branches in the `NCAR/miles-credit` Github repository. 
+
+### External contributors: Fork and clone the repository
+
+Get started by forking the `NCAR/miles-credit` repository on GitHub. To do this, find the "Fork" button near the top of the page and click it. This will create a copy of the project under your personal github account.
```
Comment (kafitzgerald) [inline docs/source/contrib.md]:
```suggestion
Get started by forking the `NCAR/miles-credit` repository on GitHub. To do this, find the "Fork" button near the top of the page and click it. This will create a copy of the project under your personal GitHub account.
```

File: docs/source/contrib.md [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+
+You can skip the pre-commit hooks by adding the `--no-verify` flag to your commit command like this:
+
+```bash  
+git commit -m "your commit message" --no-verify  
+```
+
+For more information on pre-commit hooks, see the [pre-commit documentation](https://pre-commit.com/).
+
+Pre-commit also runs through github actions as a continuous integration check through [pre-commit.ci](http://pre-commit.ci). This pre-commit will perform auto syntax checking and safe reformatting and will fail if safe reformats are not feasible.
```
Comment (kafitzgerald) [inline docs/source/contrib.md]:
```suggestion
Pre-commit also runs through GitHub Actions as a continuous integration check through [pre-commit.ci](http://pre-commit.ci). This pre-commit will perform auto syntax checking and safe reformatting and will fail if safe reformats are not feasible. 
```

Did auto-fixes get turned on?  Not opposed, but last I remember we needed a bot command the way it was configured on here.

File: docs/source/contrib.md [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+git commit -m "your commit message" --no-verify  
+```
+
+For more information on pre-commit hooks, see the [pre-commit documentation](https://pre-commit.com/).
+
+Pre-commit also runs through github actions as a continuous integration check through [pre-commit.ci](http://pre-commit.ci). This pre-commit will perform auto syntax checking and safe reformatting and will fail if safe reformats are not feasible. 
+
+## Understanding the CREDIT Codebase
+
+Please read the CREDIT User Guide to understand the structure of the code in more detail, especially sections related to what you want to work on.
```
Comment (kafitzgerald) [inline docs/source/contrib.md:168]: Can we link to the "User Guide" or a relevant section?  I'd add a suggestion, but I'm not sure what would be best.

File: docs/source/contrib.md [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+
+## Feature Integration Process
+
+CREDIT Gen 2 now allows users to create and integrate new datasets, pre and post blocks, models, and losses as custom objects without inserting them into the CREDIT codebase. Before asking to include your new code into CREDIT, first register it as a custom object in your config file.
```
Comment (kafitzgerald) [inline docs/source/contrib.md:172]: Might need more context on what Gen 2 is.  For now at least.

Or maybe there is something we can link to?

File: docs/source/contrib.md [hunk 10]
```diff
@@ -0,0 +1,347 @@
...
+   ```
+
+   ## Commit Guidelines
+
+   - Do NOT include any AI authorship, attribution, or co-author lines.
+
+   - Never append "Co-Authored-By: Claude" or "Generated by Claude" to messages.
+
+    ```
+
```
Comment (kafitzgerald) [inline docs/source/contrib.md]: I'm curious about these commit guidelines and the motivation here.  Will follow up more offline to avoid bogging this down.

File: docs/source/contrib.md [hunk 11]
```diff
@@ -0,0 +1,347 @@
...
+2. Before engaging with an LLM, write a clear, specific goal and specification for the coding task. Include specific details about what the expected functionality should be and how the functionality is to be tested.
```
Comment (kafitzgerald) [inline docs/source/contrib.md:192]: Much of this section seems more broadly relevant.  I wonder if there might be some high quality, well maintained external resources we could point to.  If not now perhaps in the future.

File: docs/source/contrib.md [hunk 12]
```diff
@@ -0,0 +1,347 @@
...
+
+These tests will also run automatically when you open a pull request using GitHub Actions and the `.github/workflows/ci.yml` file.
+
+While unit tests cover individual functions and classes, some errors arise from the closely integrated nature of auto-regressive emulators and errors that grow through time. Some errors also only occur on GPUs or through multi-node communication. We perform integration tests through the NSF NCAR CIRRUS on-premises cloud. Multi-GPU tests have to be performed manually on Derecho or other large HPC systems. See the tests/manual directory for examples of these manual multi-node tests and how to write your own if necessary.
+
+See the [pytest documentation](https://docs.pytest.org/en/stable/) for more information about writing and running tests.
+
+## Documentation
+
+`miles-credit` uses [sphinx](https://www.sphinx-doc.org/en/master/) and [ReadTheDocs](https://docs.readthedocs.io/en/stable/) to build and host the documentation.
```
Comment (kafitzgerald) [inline docs/source/contrib.md]:
```suggestion
`miles-credit` uses [Sphinx](https://www.sphinx-doc.org/en/master/) and [Read the Docs](https://docs.readthedocs.io/en/stable/) to build and host the documentation.
```

File: docs/source/contrib.md [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+To generate the documentation locally, follow the steps below.
+
+1. Activate the credit environment.  
+2. Enter the `docs` directory.  
+3. Run `make html` or to build the documentation.  
+4. Open `docs/_build/html/index.html` in your browser to view the documentation.
+
+### Check the documentation
+
+As well as checking local documentation generation, you should also check the preview documentation generated as part of a PR. To do this, scroll down to the "checks" section of the PR and click on the "Details" link next to the "docs/readthedocs.org:miles-credit" check. This will take you to the corresponding build on ReadTheDocs, where you can view the documentation built from your PR and see any warnings or errors on your build. ReadTheDocs will highlight tracked changes.
```
Comment (kafitzgerald) [inline docs/source/contrib.md]:
```suggestion
As well as checking local documentation generation, you should also check the preview documentation generated as part of a PR. To do this, scroll down to the "checks" section of the PR and click on the "Details" link next to the "docs/readthedocs.org:miles-credit" check. This will take you to the corresponding build on Read the Docs, where you can view the documentation built from your PR and see any warnings or errors on your build. Read the Docs will highlight tracked changes. 
```

File: docs/source/contrib.md [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+
+We recommend adding a short, descriptive title to your pull request. The body of the pull request should autofill with our pull request template, which has its own set of directions. Please fill out the relevant sections of the template, including adding a more detailed description of your changes. You can use an LLM to summarize your changes in markdown format and add the description here. Please review and edit for human-readability, including defining jargon. 
+
+Please also link the associated issue with the PR with the phrase `Closes \#ZZZ`. Once the PR is merged, the related issues will also close automatically. 
+
+Once you have filled out the template, click the "Create pull request" button. This will open your pull request on the CREDIT repository.
+
+If you want to receive feedback on code changes but are not ready for them to be merged in a PR, you can post a comment to a relevant issue and link your branch. You can also highlight individual code blocks and right-click to “Get permalink” that will then show up in the comment as a code snippet.
```
Comment (kafitzgerald) [inline docs/source/contrib.md:293]: We could also mention "draft" PRs here as an option if folks are open to that.

File: docs/source/contrib.md [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+Recommended
+
+- [ ] Updated public facing methods have full docstrings.  
+- [ ] Variable names balance being unambiguous and concise.  
+- [ ] Comments have been added to describe more complex operations.  
+- [ ] Code minimizes redundancy with the use of loops, function/method calls, and robust data structures.  
+- [ ] If dependencies are added, they should not burden the user with additional installation steps or break other parts of the code.  
+- [ ] If dependencies are removed, changes to existing code should be tested and any issues addressed.  
+- [ ] Tests cover expected outcomes of a function and known edge cases.  
+- [ ] Type annotation is added to inputs and outputs of functions/methods.
```
Comment (kafitzgerald) [inline docs/source/contrib.md:320]: We might consider linking to the original file or embedding that here instead so these don't get out of sync.

File: docs/source/contrib.md [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+
+After you open your pull request, the CREDIT team will review it and may provide feedback like asking for changes or suggesting improvements. You can address this feedback by making changes to your branch and pushing them to your branch. The pull request will automatically update with your changes. Please also comment with your changes. 
+
+The CREDIT team appreciates your contributions and the interactive process of reviewing pull requests, and will do our best to review your pull request in a timely manner. It is totally normal to have to make several rounds of changes to your pull request before it is ready to be merged, especially if you are new to the project.
+
+Once your pull request is approved by a core maintainer and passes the relevant checks, it will be merged into the main repository!
+
+### Delete your branch
+
+We recommend deleting your branch after your pull request is merged. This will help keep the repo clean and organized, but is not required.
```
Comment (kafitzgerald) [inline docs/source/contrib.md]: If folks are working on the main repo, a stronger recommendation might be warranted.

**Review by kafitzgerald — COMMENTED** [review]
Thanks for getting this rolling!

Looks good for the most part - some minor suggestions and comments inline and I have a couple of questions about the AI policy I'll follow up on offline.

I also noticed the Read the Docs build preview check isn't populating.  I suspect this has to do with the move to GitHub Enterprise.  One of the maintainers on there probably needs to go re-authenticate their account to get it to work again.

File: docs/source/contrib.md (hunk 8 above)
Comment (djgagne) [reply docs/source/contrib.md:168]: The user guide links will be added in a separate PR once the user guide is ready.

File: docs/source/contrib.md (hunk 15 above)
Comment (djgagne) [reply docs/source/contrib.md:320]: The original file is in a Github Actions yaml file, which makes it really difficult to link. There isn't a good way to embed markdown files in other markdown files. Will keep this separate for now.

Comment (kafitzgerald) [reply docs/source/contrib.md:320]: Ah, right.

File: docs/source/contrib.md (hunk 4 above)
Comment (kafitzgerald) [reply docs/source/contrib.md:28]: I'm going to go ahead and resolve this.  I think an issue template is a better home for this level of detail anyway.

**Review by kafitzgerald — APPROVED** [review]
